### PR TITLE
add sign functionality

### PR DESCRIPTION
**Pull request by diveshuttam**
I have changed the commands using basename as otherwise the bash process was seperate and other egpg functions like `call cmd_sign` etc. were not accessible.

There is one thing left in here, for all gui commands we need to change pinentry-tty to something like pinentry-gtk or some other graphical user interface.

### CODE REVIEW

**Comment by dashohoxha** [comment]
> I have changed the commands using basename as otherwise the bash process was seperate and other egpg functions like call cmd_sign etc. were not accessible.

:+1: I did not test it that far.

> There is one thing left in here, for all gui commands we need to change pinentry-tty to something like pinentry-gtk or some other graphical user interface.

On `cmd_init` we set the configuration of gpg-agent to `pinentry-program /usr/bin/pinentry-tty`. Maybe the initialization for the GUI case should be a bit different. If we set no value at all for pinentry, I think that the default value is `pinentry-gtk`.

By the way, we also need to do something about `gpg.conf` because it is deprecated (or useless) in gnupg-2.2.

Another thing is that on the file `config.sh` we need to save a label as well. This is useful for the GUI case but also for the command line. The purpose of this label is to help the user distinguish clearly and easily one key from another (in a scenario when he may need to use more than one key). For example he may use one key for communicating with work colleagues, another one for communicating with friends, and a third one for communicating with family members. Each of these keys lives in a different context (homedir). Normally the user should be able to distinguish them from each-other by the key id (or fingerprint). However this is more difficult and error prone than distinguishing them by a simple and user friendly label.
Maybe we should also allow the user to switch easily the context of egpg from the GUI.

**Comment by diveshuttam** [comment]
> By the way, we also need to do something about gpg.conf because it is deprecated (or useless) in gnupg-2.2.

I think I have been misinterpreting some of the things here, I again read the discussion on gnupg-devel, It says that keyserver entries are deprecated they should be shifted to dirmngr.conf instead.
So I think other things should work fine.

pinentry-program is to be given in gpg-agent.conf so there should be no problem there too. BTW I am just trying these things.

**Comment by dashohoxha** [comment]
In case that `$file.signed` already exists (it may be a fake or corrupted one) signing fails.
I think that `gpg` in this case tries to ask whether it should override it or not, but since it is not running in interactive mode it fails silently.

File: src/gui/sign.sh [hunk 1]
```diff
@@ -1,7 +1,9 @@
 gui_sign() {
-    yad --file
-    #yad --title="EasyGnuPG | Sign a File" \
-    #   --text="Sign a File"
+    local file output
+    file=$(yad --file --title="EasyGnuPG | Sign a File")
```
Comment (dashohoxha) [inline src/gui/sign.sh]: We should remove the `$file.signature` in case it exists, or ask the user first whether it should be replaced.

**Review by dashohoxha — CHANGES_REQUESTED** [review]

**Review by dashohoxha — APPROVED** [review]

**Comment by dashohoxha** [comment]
It looks OK as far as I have tested. You can merge it.

File: src/gui/sign.sh [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
+        rm -f "$file.signature"
     fi
-    
+
     output=$(call cmd_sign $file)
     [[ -n "$output" ]] && message error "$output"
-    [[ -f "$file.signature" ]] && message info "Signature saved as $file.signature"
+
+    if [[ -s "$file.signature" ]]; then
+        yad --file --filename="$file.signature" &
```
Comment (diveshuttam) [inline src/gui/sign.sh]: Hi, I didn't understand the purpose of line 15.
Is it just showing the file to the user? Or you wanted something like user being able to select the desired signature file?

Comment (dashohoxha) [reply src/gui/sign.sh]: > Is it just showing the file to the user?

Yes.

**Pull request merged**